<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2023-03-21" />

<title>All data processing</title>

<script src="site_libs/header-attrs-2.13/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">bioc 202</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="schedule.html">Schedule</a>
</li>
<li>
  <a href="syllabus.html">Syllabus</a>
</li>
<li>
  <a href="exercises.html">Exercises</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">All data processing</h1>
<h4 class="date">2023-03-21</h4>

</div>


<style>
.text-box {
  background-color: #d4e9fc;
  color: black;
  font-size: 14px;
  border-radius: 5px; 
  padding: 20px
}
</style>
<style>
.lecture-box {
  background-color: #f7e1fc;
  color: black;
  font-size: 14px;
  border-radius: 5px; 
  padding: 20px
}
</style>
<div id="processing-data" class="section level1">
<h1>processing data</h1>
<p>When the data comes back from the sequencer, it is one giant text
file in fastq format that includes all data from your lane of
sequencing. Remember about barcodes- we can use these to demultiplex our
data (or really the sequencing center will do it for you, depending on
your data type).</p>
<p>In general, you’ll take some fairly consistent steps, with a lot of
options, to process your data so you can get to the point of calling
snps. These are: quality control, cleaning, and mapping.</p>
<p>For our purposes will use just one sample to learn about data
processing. This is sample, <code>TR-023</code>, which is randomly
chosen. After demultiplexing, we have two files associated with this
sample: <code>TR-023_RA.fastq.gz</code> and
<code>TR-023_RB.fastq.gz</code></p>
<div class="text-box">
<p>Why do we have two files for one sample?</p>
</div>
<p>example of fastq format:</p>
<pre><code>@J00113:190:HFV3LBBXX:7:1101:3742:1297 1:N:0:NAGCTT
TGCAGGCTCAGTCCTGATGAGGGAGCCATCTTTATAGAAAGCAGCTGGGAGGTTGGAGGGAGCGGTCTTTGTGTGACAGCTCAGAGTGACGGCAGCTCCCTCCATCACATGGAGTACAGGATTCTGCATGATTTCTGATC
+
FJJJJFJJJFJFJJJJAJJ-FJJFFJJJAAAAF-&lt;AJFFFFJJJFAFFF&lt;A7-&lt;FFAJJJJ&lt;JAJ7A&lt;AJFJJJJJJJJJA-FAJA-AAFF&lt;AF-F--AF7FAJJ77-7JJJA-7--AJJA-AFFF-----&lt;--7--777
@J00113:190:HFV3LBBXX:7:1101:13738:1314 1:N:0:NAGCTT
TGCAGGACTCGTTGTTGATGAGGAAATGGTCACCACATGCTGACATTGTGGGTAAGCCAGTATTTCAAATAGTGGTTCCCAGAAAACTCCGTTCTGCAGTGTGAAAAGTTGCTCTTGATGAGTCTGGTCATTCTGGTGTA
+
JJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJJ&lt;FJ7FJJJFJJJJJJJFJJJJJJ-&lt;FFFJJJ-A-F)7&lt;J-7&lt;-&lt;F-AFJ
@J00113:190:HFV3LBBXX:7:1101:19015:1314 1:N:0:NAGCTT
TGCAGGTTTAAGCTCAGAAGACGACGCCCCTGCACTGAAATACAGGAAGAAGATCTGCATGAACCAGCCTCACCTGCGGTTTGAAATCACCAGCGAGGACGGCTTCAGCGTTAAGGCTAACAGCATAGAGGGTAGGAGTG
</code></pre>
<p>In the fastq file, each read consists of 4 lines:</p>
<ol style="list-style-type: decimal">
<li>A sequence identifier with information about the sequencing run and
the sequence</li>
<li>The actual sequence</li>
<li>A separator (+)</li>
<li>the quality scores.</li>
</ol>
<p>You can look at your fastq file yourself. However, these are gz
files. Try to look at them with <code>head</code>.</p>
<p>You need to use <code>zcat your_fastqFile</code>. Warning! This will
print your whole file! You can do this and see what happens.
<code>control + c</code> will cancel the printing.</p>
<p>Instead, we need to use <code>|</code>, or a pipe. So we can pipe the
output of <code>zcat</code> to another command, in this case
<code>head</code>.</p>
<p><code>zcat TR-023_RA.fastq.gz  | head</code></p>
<p>We can also count how many lines are in our file. you do this with
<code>wc -l</code>.</p>
<div class="text-box">
<p>Use <code>zcat</code>, <code>|</code>, and <code>wc -l</code> to
count the number of reads in your fastq files.</p>
<ul>
<li>How many reads do you have?<br />
</li>
<li>Do the number of forward and reverse reads agree? Why?</li>
</ul>
</div>
<div id="check-quality" class="section level2">
<h2>check quality</h2>
<p>We will use <a
href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">fastqc</a>
to check the quality of our sequencing. Briefly, this program scans our
fastq files and checks for quality, contamination from adapters, and
sequence bias, among others.</p>
<p>Look at the help page of fastqc using <code>fastqc --help</code></p>
<p>We can then run fastqc for our forward read:</p>
<pre class="bash"><code>fastqc course_materials/TR-023_RA.fastq.gz -o .</code></pre>
<div class="text-box">
<ul>
<li>Modify the above code to run fastqc on your reverse read.</li>
<li>Discuss the output of fastqc together.</li>
</ul>
</div>
</div>
<div id="trim" class="section level2">
<h2>trim</h2>
<p>We don’t have much evidence for adapter contamination with this
sample, but we do see evidence for some quality issues, which are
normal. We can deal with these issues by trimming. Note, this isn’t
totally necessary in all cases as many alignment software can deal with
low quality bases.</p>
<p>We will use <a
href="https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/">Trim
Galore</a> to clean up our data. There are many programs you can use for
this- we’re using Trim Galore because it is relatively simple and
because it auto detects adapters.</p>
<p>Below is the code to run this program. Again, you can get help using
<code>trim_galore --help</code></p>
<pre class="bash"><code>trim_galore --paired \
      course_materials/TR-023_RA.fastq.gz \
      course_materials/TR-023_RB.fastq.gz \
      -q 30 --length 30 --stringency 4</code></pre>
<div class="text-box">
<p>After trim_galore finishes, run fastqc on your newly trimmed
data.</p>
<ul>
<li>What has changed? Anything?</li>
<li>Change the stringency, quality, or length threshold in your
trim_galore command. How has this affected your results?</li>
</ul>
</div>
</div>
<div id="align-to-reference-genome" class="section level2">
<h2>align to reference genome</h2>
<p>We’re not actually doing this!</p>
<div class="lecture-box">
<p>Short lecture on alignment, reference genomes, and Stacks.</p>
</div>
<div id="understanding-mapping-output" class="section level3">
<h3>Understanding mapping output</h3>
<p>Output files are in <em>sam</em> or <em>bam</em> format where bam
files are compressed versions of sam files.</p>
<div class="text-box">
<p>Use <code>head</code> to look at your bam output. What do you
see?</p>
</div>
<p>We need to use another program, <code>samtools</code> to look at our
bam file. Again, use <code>samtools --help</code> or just
<code>samtools</code> to get an overview of this program.</p>
<p>The most basic tool in samtools is <code>samtools view</code>, which
prints the bam file in plain text. You can see the help page with
<code>samtools view --help</code>.</p>
<p>the basic syntax is: <code>samtools view NAME.bam</code>…. remeber
<code>head</code> and <code>|</code>.</p>
<div class="text-box">
<p>Look at your bam file using samtools.</p>
</div>
<p>You should see something like this:</p>
<pre><code>J00113:190:HFV3LBBXX:7:1223:20912:25914 163     NW_012224401.1  27218   60      150M    =       27487   409     CACACAGCTGCTGAGTAGCCAGAATATGAAAAGCAGTGTCAAGCTGTTCATATCTGGGTAAAGTTTTGAGTTTGGAAACGTATCTACAGCTGATATTACATGGTTATAATACACTGATTGATGAGCGTCTTTTCTCTGACTCATCCTGGT  AAFFFJJJJJJJJJJFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJAJJJJJJJJJJJJJJJJJFJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJAFJJJJJFFJF&lt;FFJJJJJJJJJJJJJJJJJ      NM:i:1  MD:Z:60G89      AS:i:145        XS:i:115    RG:Z:AC-3-TR-023-023
</code></pre>
<p>So what’s going on in this file? Its a lot to digest.</p>
<table>
<colgroup>
<col width="3%" />
<col width="12%" />
<col width="49%" />
<col width="34%" />
</colgroup>
<thead>
<tr class="header">
<th align="right">Col</th>
<th align="left">Field</th>
<th align="left">Brief_description</th>
<th align="left">our_read</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">1</td>
<td align="left">QNAME</td>
<td align="left">Query template NAME</td>
<td align="left">J00113:190:HFV3LBBXX:7:1223:20912:25914</td>
</tr>
<tr class="even">
<td align="right">2</td>
<td align="left">FLAG</td>
<td align="left">bitwise FLAG</td>
<td align="left">163</td>
</tr>
<tr class="odd">
<td align="right">3</td>
<td align="left">RNAME</td>
<td align="left">References sequence NAME</td>
<td align="left">NW_012224401.1</td>
</tr>
<tr class="even">
<td align="right">4</td>
<td align="left">POS</td>
<td align="left">1-based leftmost mapping POSition</td>
<td align="left">27218</td>
</tr>
<tr class="odd">
<td align="right">5</td>
<td align="left">MAPQ</td>
<td align="left">MAPping Quality</td>
<td align="left">60</td>
</tr>
<tr class="even">
<td align="right">6</td>
<td align="left">CIGAR</td>
<td align="left">CIGAR string</td>
<td align="left">150M</td>
</tr>
<tr class="odd">
<td align="right">7</td>
<td align="left">RNEXT</td>
<td align="left">Ref. name of the mate/next read</td>
<td align="left">=</td>
</tr>
<tr class="even">
<td align="right">8</td>
<td align="left">PNEXT</td>
<td align="left">Position of the mate/next read</td>
<td align="left">27487</td>
</tr>
<tr class="odd">
<td align="right">9</td>
<td align="left">TLEN</td>
<td align="left">observed Template LENgth</td>
<td align="left">409</td>
</tr>
<tr class="even">
<td align="right">10</td>
<td align="left">SEQ</td>
<td align="left">segment SEQuence</td>
<td align="left">CACACAG…</td>
</tr>
<tr class="odd">
<td align="right">11</td>
<td align="left">QUAL</td>
<td align="left">ASCII of Phred-scaled base QUALity plus 33</td>
<td align="left">AAFFFFJ…</td>
</tr>
<tr class="even">
<td align="right">12</td>
<td align="left">NM:i:count</td>
<td align="left">Number of differences between the sequence and
reference</td>
<td align="left">NM:i:1</td>
</tr>
<tr class="odd">
<td align="right">13</td>
<td align="left">MD:Z:</td>
<td align="left">String encoding mismatched and deleted reference
bases</td>
<td align="left">MD:Z:60G89</td>
</tr>
<tr class="even">
<td align="right">14</td>
<td align="left">AS:i:score</td>
<td align="left">Alignment score generated by aligner</td>
<td align="left">AS:i:145</td>
</tr>
<tr class="odd">
<td align="right">15</td>
<td align="left">XS:i:score</td>
<td align="left">Seconday alignment score</td>
<td align="left">XS:i:115</td>
</tr>
<tr class="even">
<td align="right">16</td>
<td align="left">RG:Z:readgroup</td>
<td align="left">Read group identification</td>
<td align="left">RG:Z:AC-3-TR-023-023</td>
</tr>
</tbody>
</table>
</div>
<div id="overall-alignment-statistics" class="section level3">
<h3>overall alignment statistics</h3>
<p>The easiest way to look at how well our mapping worked is to use
<code>samtools flagstat TR-023.bam</code>, which gives us this
output:</p>
<pre><code>1147073 + 0 in total (QC-passed reads + QC-failed reads)
1124360 + 0 primary
0 + 0 secondary
22713 + 0 supplementary
238412 + 0 duplicates
234206 + 0 primary duplicates
1127675 + 0 mapped (98.31% : N/A)
1104962 + 0 primary mapped (98.27% : N/A)
1124360 + 0 paired in sequencing
562180 + 0 read1
562180 + 0 read2
1003448 + 0 properly paired (89.25% : N/A)
1097766 + 0 with itself and mate mapped
7196 + 0 singletons (0.64% : N/A)
86662 + 0 with mate mapped to a different chr
53567 + 0 with mate mapped to a different chr (mapQ&gt;=5)</code></pre>
<div class="lecture-box">
<p>Short lecture/discussion on these results</p>
<p><a
href="https://www.molecularecologist.com/2016/08/25/the-trouble-with-pcr-duplicates/"
class="uri">https://www.molecularecologist.com/2016/08/25/the-trouble-with-pcr-duplicates/</a></p>
<p><a
href="https://hbctraining.github.io/Intro-to-rnaseq-hpc-O2/lessons/04_alignment_quality.html"
class="uri">https://hbctraining.github.io/Intro-to-rnaseq-hpc-O2/lessons/04_alignment_quality.html</a></p>
</div>
<div id="a-more-in-depth-look" class="section level4">
<h4>A more in-depth look</h4>
<p>Remember the flags in your bam file? You can filter your reads based
on this (which is what <code>samtools flagstat</code> is doing).
However, we can be more specific. The syntax to do this is:
<code>samtools view -f # TR-023.bam</code>. Where <code>-f</code> can
also be <code>-F</code> and <code>#</code> is the flag you want to
use.</p>
<p><code>-f</code> to find the reads that agree with the flag statement
<code>-F</code> to find the reads that do not agree with the flag
statement</p>
<p>At <a
href="http://broadinstitute.github.io/picard/explain-flags.html">this
website</a> you can generate any flag definition you want (or get a flag
explained).</p>
<div class="text-box">
<p>Go to <a
href="http://broadinstitute.github.io/picard/explain-flags.html"
class="uri">http://broadinstitute.github.io/picard/explain-flags.html</a></p>
<p>How would you find the number of mapped reads in your bam file? Does
this number agree with <code>samtools flagstat</code>?</p>
<p>You can also filter by quality by adding a <code>-q 20</code> option.
A score of 20 is a common threshold. Count the number of mapped reads
with q of 20. What do you find?</p>
<p>We can also look at the mapping quality overall:</p>
<pre><code>samtools view TR-023.bam | cut -f 5 &gt; mapq.txt</code></pre>
<p>Plot a histogram of these mapping qualities in R (hint, just use
<code>hist()</code>).</p>
</div>
<p><a
href="https://www.slideshare.net/lindenb/ngsformats?ref=http://plindenbaum.blogspot.com/2013/09/presentation-file-formats-for-next.html?m=1"
class="uri">https://www.slideshare.net/lindenb/ngsformats?ref=http://plindenbaum.blogspot.com/2013/09/presentation-file-formats-for-next.html?m=1</a></p>
<div id="call-snps" class="section level5">
<h5>call snps</h5>
<p>Not doing this!</p>
<div class="lecture-box">
<p>Short snp calling lecture</p>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="working-with-snp-data" class="section level1">
<h1>working with SNP data</h1>
<p>The most common file format for SNPs is VCF.</p>
<p>~/admixture_mapping/variants/all.chrom.vcf.gz</p>
<p>vcftools</p>
<div id="filter-snps" class="section level2">
<h2>filter snps</h2>
<p><a href="https://www.ddocent.com/filtering/"
class="uri">https://www.ddocent.com/filtering/</a> <a
href="https://evomics.org/wp-content/uploads/2022/06/Population-Genomics-Lab-1.pdf"
class="uri">https://evomics.org/wp-content/uploads/2022/06/Population-Genomics-Lab-1.pdf</a></p>
<div id="pca" class="section level3">
<h3>pca</h3>
<p>add snp ids:</p>
<p>chr1_29988</p>
<p>need to get format right. its annoying.</p>
<pre class="bash"><code>
zcat variants.vcf.gz | grep -v &#39;^#&#39; &gt; variants_noheader.vcf

grep -v &quot;NW_&quot; variants_noheader.vcf &gt; variants_noheader.vcf2
grep -Ew &quot;chr1|chr2|chr3|chr4|chr5&quot; variants_noheader.vcf2 &gt; variants_noheader.vcf3

mv variants_noheader.vcf3 variants_noheader.vcf
rm variants_noheader.vcf2

zcat variants.vcf.gz | grep &#39;^#&#39; &gt; header.txt

cut -f 1 variants_noheader.vcf  &gt; chr.txt
cut -f 2 variants_noheader.vcf | paste chr.txt - -d &quot;:&quot; &gt; snp_id.txt
cut -f 1-2 variants_noheader.vcf | paste - snp_id.txt &gt; threecols.txt
cut -f 4- variants_noheader.vcf | paste threecols.txt - | cat header.txt - &gt; all.vcf



</code></pre>
<pre class="bash"><code>/plink --vcf all.vcf --out variants_pruned \
--indep-pairwise 50 5 0.2 --allow-extra-chr --double-id</code></pre>
<p>REWRITE THIS: This command is removing SNPs in linkage using the
-–indep-pairwise option. The numbers after the option control how plink
removes these SNPs. The first number is the window size (in the number
of variants it investigates). The second number is the frame-shift for
the window size (how many variants the window will move between
analysis). The third number is the r2 cut-off value (designating an
upper limit for how correlated/in-linkage SNPs in a window are allowed
to be). 50 5 and 0.2 are commonly used values</p>
<p>make the pruned file:</p>
<pre class="bash"><code>/plink --vcf all.vcf \
--extract variants_pruned.prune.in \
--make-bed --out variants_NoLD \
--allow-extra-chr --double-id</code></pre>
<p>run the pca</p>
<pre class="bash"><code>/plink --bfile variants_NoLD \
--pca --out variants_NoLD_PCA --allow-extra-chr --double-id</code></pre>
<p>plot the pca results</p>
<pre class="r"><code>library(ggplot2)

setwd(&quot;~/Documents/GEOMAR/Teaching/Mar_pop_gen/2023/teaching&quot;)

dat &lt;- read.table(&quot;variants_NoLD_PCA.eigenvec&quot;, header=F)

colnames(dat) &lt;- c(&quot;ID&quot;, &quot;ID2&quot;, &quot;PC1&quot;, &quot;PC2&quot;, &quot;PC3&quot;, &quot;PC4&quot;, colnames(dat)[7:ncol(dat)])

dat$population &lt;- substr(dat$ID, 1,2)

d &lt;- ggplot(dat, aes(PC1, PC2, fill=population)) +
        geom_point(size=4.5, shape=21, color=&quot;black&quot;) +
        #xlab(paste0(&quot;PC1: &quot;,percentVar[1],&quot;% variance&quot;)) +
        #ylab(paste0(&quot;PC2: &quot;,percentVar[2],&quot;% variance&quot;)) +
        theme_bw() +
        #scale_fill_manual(values=c(&#39;steelblue1&#39;,&#39;steelblue&#39;,&#39;grey45&#39;, &quot;darkorchid2&quot;, &quot;firebrick3&quot;),
  scale_fill_manual(values=c(&quot;#D3DDDC&quot;,&#39;#6699CC&#39;,&quot;#F2AD00&quot;,&quot;#00A08A&quot;, &quot;#CC3333&quot;),
                    labels = c(&quot;Founding population&quot;,&quot;Ambient&quot;, &quot;Acidic&quot;, 
                               &quot;Warming&quot;, &quot;Greenhouse&quot;))+
        #theme(legend.title = element_blank())+
       # theme(legend.text=element_text(size=8))+
        #theme(legend.background = element_rect(colour = &#39;black&#39;, fill = &#39;white&#39;, linetype=&#39;solid&#39;))+
#        ggtitle(&quot;F1&quot;)+
        #fill=c(&#39;steelblue1&#39;,&#39;steelblue&#39;,&#39;grey45&#39;, &quot;darkorchid2&quot;, &quot;firebrick3&quot; )),order = 2),
    fill=c(&quot;#D3DDDC&quot;,&#39;#6699CC&#39;,&quot;#F2AD00&quot;,&quot;#00A08A&quot;, &quot;#CC3333&quot;)),order = 2),
    shape= FALSE)

#ggsave(&quot;~/tonsa_genomics/figures/pca_afs_noF3.pdf&quot;,d, w=5.5, h=3.7)</code></pre>
</div>
<div id="admixture" class="section level3">
<h3>admixture</h3>
<p>short lecture on population structure, Fst, drift.</p>
<pre class="bash"><code>
for K in 1 2 3 4 5 6; \
do admixture --cv variants_NoLD.bed $K | tee log${K}.out; done
</code></pre>
<pre class="r"><code>samplelist &lt;- read_tsv(&quot;~/Documents/GEOMAR/Teaching/Mar_pop_gen/2023/teaching/indivs_subsample.txt&quot;,
                       col_names = &quot;sample&quot;)

read_delim(&quot;~/Documents/GEOMAR/Teaching/Mar_pop_gen/2023/teaching/variants_NoLD.2.Q&quot;,
                  col_names = paste0(&quot;Q&quot;,seq(1:2)),
                  delim=&quot; &quot;)


# read in all date, in a loop
all_data &lt;- tibble(sample=character(),
                   k=numeric(),
                   Q=character(),
                   value=numeric())

head(all_data)

for (k in 1:6){
  data &lt;- read_delim(paste0(&quot;~/Documents/GEOMAR/Teaching/Mar_pop_gen/2023/teaching/variants_NoLD.&quot;,k,&quot;.Q&quot;),
                  col_names = paste0(&quot;Q&quot;,seq(1:k)),
                  delim=&quot; &quot;)
  data$sample &lt;- samplelist$sample
  data$k &lt;- k
  #This step converts from wide to long.
  data %&gt;% gather(Q, value, -sample,-k) -&gt; data
  all_data &lt;- rbind(all_data,data)
}

# add the population label
all_data$population &lt;- substr(all_data$sample, 1, 2)
all_data$population &lt;- factor(all_data$population, 
                              levels=c(&quot;GA&quot;, &quot;PL&quot;, &quot;HP&quot;, &quot;BC&quot;, &quot;PC&quot;, &quot;TR&quot;))

# our orders are off in our vcf. lets re-order these from south to north. 
orderlist &lt;- read_tsv(&quot;~/Documents/GEOMAR/Teaching/Mar_pop_gen/2023/teaching/population_order.txt&quot;,
                       col_names = &quot;sample&quot;)
all_data$sample&lt;-factor(all_data$sample,levels=orderlist$sample)

all_data %&gt;%
  filter(k == 2) %&gt;%
  ggplot(.,aes(x=sample,y=value,fill=factor(Q))) + 
  geom_rug(aes(x=sample, y=value, color=population)) +
  geom_bar(stat=&quot;identity&quot;,position=&quot;stack&quot;) +
  xlab(&quot;Sample&quot;) + ylab(&quot;Ancestry&quot;) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_fill_brewer(palette=&quot;Set1&quot;,name=&quot;K&quot;,
                    labels=c(&quot;1&quot;,&quot;2&quot;))

# plot all k values.
p &lt;-  ggplot(all_data,aes(x=sample,y=value,fill=factor(Q))) + 
  geom_bar(stat=&quot;identity&quot;,position=&quot;stack&quot;) +
  geom_rug(aes(x=sample, color=population), inherit.aes=F) +
  xlab(&quot;Sample&quot;) + ylab(&quot;Ancestry&quot;) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 60, hjust = 1)) +
  scale_fill_brewer(palette=&quot;Set1&quot;,name=&quot;K&quot;,
                    labels=seq(1:5)) +
  facet_wrap(~k,ncol=1)



ggsave(&quot;Admixture_plot.pdf&quot;, p, width = 7, height = 15, units=&quot;in&quot;)</code></pre>
</div>
<div id="fst" class="section level3">
<h3>fst</h3>
<p>can do big windows, then selection scan?</p>
<p><a href="https://owensgl.github.io/biol525D/Topic_8-9/fst.html"
class="uri">https://owensgl.github.io/biol525D/Topic_8-9/fst.html</a></p>
<pre class="bash"><code>
vcftools --gzvcf all.vcf \
--weir-fst-pop TR.pop \
--weir-fst-pop GA.pop \
--fst-window-size 25000 --out TR_GA_25kb
</code></pre>
<p>plot the pairwise fst density plots</p>
<pre class="r fold-hide"><code># read in all in a list

take all, merge, make ridgeplot for each.
a
a
a
a
a</code></pre>
<p>plot correlation plots</p>
</div>
<div id="diversity" class="section level3">
<h3>diversity</h3>
<p>these run decently fast.</p>
<pre class="bash"><code>for pop in TR PC BC HP PL GA;
do
  vcftools --vcf all.vcf \
    --keep ${pop}.pop \
    --window-pi 50000 \
    --out ${pop}_pi_50kb;
done</code></pre>
<pre class="r"><code># read in all in a list

#take all, merge, make ridgeplot for each. </code></pre>
</div>
</div>
<div id="gene-environment-associations-or-selection-scans."
class="section level2">
<h2>gene environment associations or selection scans.</h2>
<p>baypass? bayenv? lfmm? pcadapt is probably easiest.</p>
</div>
<div id="gwas" class="section level2">
<h2>gwas</h2>
<p>gemma?</p>
</div>
<div id="intersection-between-the-two" class="section level2">
<h2>intersection between the two?</h2>
</div>
</div>
<div id="things-to-do" class="section level1">
<h1>things to do:</h1>
<p>consider parsing down to fewer chromosomes</p>
<p>Asmixture runs too slow. drop some samples?</p>
<p>R libraries: ggplot2 ggridges</p>
<p>drop the funny sample PC-393-024 Probably drop PL- it is
confusing.</p>
<p>fix the error in the vcf header that gets thrown with VCFtools.</p>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
