---
title: "Class 1"
output:
  prettydoc::html_pretty:
    theme: cayman
    highlight: github
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(kableExtra)
library(knitr)
library(dplyr)
library(tidyr)  
```

<style>
.text-box {
  background-color: #d4e9fc;
  color: black;
  font-size: 14px;
  border-radius: 5px; 
  padding: 20px
}
</style>


# Format of practical

Tentative schedule:

```{r echo = FALSE, results = 'asis',warning=FALSE,message=FALSE}
#library(pander)
mytable = read.csv('../schedule.csv',header=T)
options(knitr.kable.NA = '')

kbl(mytable) %>%
  kable_styling(bootstrap_options = "striped", full_width = F, position = "center",
                fixed_thead = T) %>% 
  column_spec(1, bold=F, width = "2em") %>%
  column_spec(2, bold=T, width = "8em") %>%
  column_spec(3, width = "20em") %>%
  column_spec(4, width = "20em") %>%
  column_spec(5, width = "20em") %>%
  column_spec(6, width = "20em") %>%
  row_spec(seq(1,10,2), bold = F, color = "black", background = "#FFF3C0") %>%
  row_spec(0, color = "black")

```

# Study system we'll be focusing on

Our study system will be the killifish, [Fundulus heteroclitus](https://www.dnr.sc.gov/marine/mrri/acechar/speciesgallery/Fish/Mummichog/index.html#:~:text=Mummichogs%2C%20locally%20called%20mud%20minnows,and%20other%20shallow%20coastal%20habitats.) (also called mummichog or mud minnow). 

This species is the best for lots of reasons. They're broadly distributed along the east coast of North America, from Florida into Canada. 

![distribution of F. heteroclitus (map credit: Alcaraz-Hernandez and Garcia-Berthou)](images/distribution.png){width=400px}

They're the most abundant vertebrate in the habitats where they live and, given their broad range, they're super tolerant of lots of conditions: broad temperatures, salinity, hypoxia, among others. They also don't move much, with an adult typically staying in the marsh where it is born and not moving more than a few hundred meters over its life. 

<div class="text-box">

Given the high population sizes and low migration, what might we predict about populations of killifish from different regions of their distribution?

</div>

## study system

We're going to work with data from a 2018 paper in Global Change Biology, led by Tim Healy [link to paper](https://doi.org/10.1111/gcb.14386). I encourage you NOT to look at the publication, as we're going to go through a bunch of these results together for this tutorial. 

We sampled fish from across the range of the eastern United States, about 30 fish per site. 
```{r, echo=FALSE}
# create table with kable
#kable(df, 
#      col.names = c("Population Location", "Population ID", "N"), 
#      align = c("l", "l", "r"))


# Generate example table
df <- data.frame(
  population_location = c("GA", "S.NJ", "C.NJ", "N.NJ", "NH"),
  population_ID = c("GA", "HP", "BC", "PC", "TR"),
  N = c(25, 46, 50, 44, 45)
)

# Create table with kableExtra
#kbl(df,
#    col.names = c("Population Location", "Population ID", "N")) %>%
#  kable_styling(bootstrap_options = "striped", full_width = F, position = "left",
#                fixed_thead = T, latex_options = c("scale_down", "resizebox=0.7\\textwidth"))

kable(df, 
      col.names = c("Population Location", "Population ID", "N"), 
      align = c("c", "c", "c")) %>%
    kable_styling(full_width = F)



  
```

Below is a map of the sampling locations. We knew about their population structure ahead of time, so if you're thinking the sampling looks funny, it will become clear why we did this as we analyze the data together. 


<center>

![Map of sampling locations](images/map.png){width=400px}
</center>


There are also a lot of interesting phenotypic differences between populations with higher temperature tolerance and hypoxia tolerance in southern populations relative to northern populations.

<center>
![Phenotypic differences between populations](images/phenotypes.png){width=400px}
</center>

Given this, there is evidence for local adaptation and we might expect to see population structure. In the tutorial the rest of the week, you're going to analyze this dataset to identify patterns of population structure and signals of selection.

### Genomic data collection

For sequencing, we used RAD-seq with SbfI, an 8 cutter. These were sequenced on an Illumina HiSeq 4000 with 150 bp paired end reads. 





